File: CEopt-1.0/MainCEoptExample2Ext.m
% -----------------------------------------------------------------
%  MainCEoptExample2Ext.m
% -----------------------------------------------------------------
%  programmer: Americo Cunha
%              americo.cunhajr@gmail.com
%
%  Originally programmed in: Jun 18, 2021
%           Last updated in: Jul 24, 2024
% -----------------------------------------------------------------
%  ﻿Example 2: The Peaks function in 2D
% -----------------------------------------------------------------

clc; clear; close all;

disp(' ------------------- ')
disp(' MainCEoptExample2.m ')
disp(' ------------------- ')

% random number generator (fix the seed for reproducibility)
rng_stream = RandStream('mt19937ar','Seed',30081984);
RandStream.setGlobalStream(rng_stream);

% objective function
F = @(x)PeaksFunc(x);

% bound for design variables
lb = [-3; -3];
ub = [ 3;  3];

% initialize mean and std. dev. vectors
mu0    = lb + (ub-lb).*rand(2,1);
sigma0 = 5*(ub-lb);

% define parameters for the CE optimizer
CEstr.isVectorized = 1;       % Vectorized function
CEstr.EliteFactor  = 0.1;     % Elite samples percentage
CEstr.Nsamp        = 50;      % Number of samples
CEstr.MaxIter      = 80;      % Maximum number of iterations
CEstr.TolAbs       = 1.0e-2;  % Absolute tolerance
CEstr.TolRel       = 1.0e-2;  % Relative tolerance
CEstr.alpha        = 0.7;     % Smoothing parameter
CEstr.beta         = 0.8;     % Smoothing parameter
CEstr.q            = 10;      % Smoothing parameter

% CE optimizer
tic
[Xopt, Fopt, ExitFlag, CEstr] = CEopt(F, mu0, sigma0, lb, ub, [], CEstr)
toc

% meshgrid for visualization
Ngrid  = 100;
xRange = linspace(lb(1),ub(1),Ngrid);
yRange = linspace(lb(2),ub(2),Ngrid);
[X, Y] = meshgrid(xRange, yRange);
Z = F([X(:), Y(:)]);
Z = reshape(Z, size(X));

% custom colors
MyYellow = [0.9290 0.6940 0.1250];
MyBlue   = [0.0000 0.4470 0.7410];
MyRed    = [0.6350 0.0780 0.1840];

% custom colormap
N = 50;
MyColorMap1 = viridis(N);

% set up the figure
figure;
hold on;

% plotting F(x)
[c1,fig1] = contourf(X, Y, Z, N);
colormap(MyColorMap1);
colorbar;

% highlighting the optimal point
fig2 = plot(0.22,-1.62,'o','Color',MyYellow,...
                          'MarkerSize', 15,...
                          'MarkerFaceColor',MyYellow,...
                          'DisplayName', 'Analytical Solution');

% plot initial approximation
fig3 = plot(mu0(1),mu0(2),'d', ...
                          'Color',MyRed,...
                          'LineWidth' , 1,...
                          'MarkerSize', 7,...
                          'MarkerFaceColor',MyRed,...
                          'DisplayName', 'Initial Approximation');

% plot CE iterations
for j=1:CEstr.iter
    fig4 = plot(CEstr.xmean(j,1), CEstr.xmean(j,2),'d', ...
                                                   'Color',MyRed, ...
                                                   'LineWidth',1,...
                                                   'MarkerSize',7, ...
                                                   'DisplayName','CE Iterations');
end

% plot CE solution
fig5 = plot(Xopt(1), Xopt(2),'x','Color',MyBlue,...
                             'LineWidth' , 3,...
                             'MarkerSize', 25,...
                             'DisplayName','CE Optimum');

% labeling
xlabel('x_1', 'FontSize', 20, 'FontName', 'Helvetica');
ylabel('x_2', 'FontSize', 20, 'FontName', 'Helvetica');

% legend
leg = [fig2; fig3; fig4; fig5];
leg = legend(leg,'Location','NorthEast','FontSize',12);

% Set font and box
set(gca, 'FontName', 'Helvetica');
set(gca, 'FontSize', 18);
box on;
hold off;

% save the figure
exportgraphics(gca, 'CEoptExample2.eps', 'Resolution', 300);

% Objective function
function F = PeaksFunc(x)
    x1 = x(:,1);
    x2 = x(:,2);
    F = 3*(1-x1).^2.*exp(-x1.^2 - (x2+1).^2) ...
      - 10*(x1/5 - x1.^3 - x2.^5).*exp(-x1.^2 - x2.^2)...
      - (1/3)*exp(-(x1+1).^2 - x2.^2);
end

File: CEopt-1.0-Python/MainCEoptExample2Ext.py
# -----------------------------------------------------------------
#  MainCEoptExample2Ext.py
# -----------------------------------------------------------------
#  programmer: Americo Cunha
#              americo.cunhajr@gmail.com
# 
#  Originally programmed in: Mar 27, 2025
#           Last updated in: Mar 28, 2025
# -----------------------------------------------------------------
#  Example 2: The Peaks function in 2D
# -----------------------------------------------------------------

import time
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import colormaps
from CEopt import CEopt  # Ensure CEopt.py is in the same directory

# --- Set random seed for reproducibility ---
np.random.seed(30081984)

# Close all figures
plt.close('all')

print(' ------------------- ')
print(' MainCEoptExample2Ext.py ')
print(' ------------------- ')

# -----------------------------------------------------------------
# Objective function
# -----------------------------------------------------------------
def PeaksFunc(x):
    x = np.atleast_2d(x)
    x1 = x[:, 0]
    x2 = x[:, 1]

    nSamples = x.shape[0]
    F        = np.zeros((nSamples, 1))

    for i in range(nSamples):
        F[i] = (3*(1 - x1[i])**2 * np.exp(-x1[i]**2 - (x2[i] + 1)**2) -
             10*(x1[i]/5 - x1[i]**3 - x2[i]**5) * np.exp(-x1[i]**2 - x2[i]**2) -
             (1/3)*np.exp(-(x1[i]+1)**2 - x2[i]**2))

    return F
# -----------------------------------------------------------------

# Fix the random seed for reproducibility (similar to RandStream in Matlab)
np.random.seed(30081984)

# Define objective function handle
F = PeaksFunc

# Bound for design variables (2D)
lb = np.array([-3, -3])
ub = np.array([ 3,  3])

# Initialize mean and std. dev. vectors
# mu0: randomly initialized within the bounds (as a 2-element array)
mu0    = lb + (ub - lb) * np.random.rand(2)
sigma0 = 5 * (ub - lb)

# Define parameters for the CE optimizer in a dictionary
CEstr = {}
CEstr['isVectorized'] = True        # Vectorized function
CEstr['EliteFactor' ] = 0.1         # Elite samples percentage
CEstr['Nsamp'       ] = 50          # Number of samples
CEstr['MaxIter'     ] = 80          # Maximum number of iterations
CEstr['TolAbs'      ] = 1.0e-2      # Absolute tolerance
CEstr['TolRel'      ] = 1.0e-2      # Relative tolerance
CEstr['alpha'       ] = 0.7         # Smoothing parameter for mean update
CEstr['beta'        ] = 0.8         # Smoothing parameter for std. dev. update
CEstr['q'           ] = 10          # Exponent for dynamic smoothing parameter

# Run the CE optimizer and time the execution
start_time = time.time()
Xopt, Fopt, ExitFlag, CEstr = CEopt(F , mu0, sigma0, lb, ub, None, CEstr)
elapsed_time = time.time() - start_time

print("\nCE optimizer results:")
print("Xopt =", Xopt)
print("Fopt =", Fopt)
print("ExitFlag =", ExitFlag)
print("Elapsed time: {:.4f} seconds".format(elapsed_time))

# Meshgrid for visualization
Ngrid  = 100
xRange = np.linspace(lb[0], ub[0], Ngrid)
yRange = np.linspace(lb[1], ub[1], Ngrid)
X, Y   = np.meshgrid(xRange, yRange)
pts    = np.column_stack((X.ravel(), Y.ravel()))
Z      = F(pts)
Z      = Z.reshape(X.shape)

# Custom colors
MyYellow = [0.9290, 0.6940, 0.1250]
MyBlue   = [0.0000, 0.4470, 0.7410]
MyRed    = [0.6350, 0.0780, 0.1840]

# Custom colormap (using 'viridis' with 50 discrete colors)
Ncolors = 50
MyColorMap1 = colormaps['viridis'].resampled(Ncolors)

# Set up the figure for visualization
plt.figure()
plt.clf()

# Plot F(x) as a filled contour plot
cont = plt.contourf(X, Y, Z, Ncolors, cmap=MyColorMap1)
plt.colorbar()

# Plot the analytical solution
plt.plot(0.22, -1.62, 'o', color=MyYellow, markersize=15,
         markerfacecolor=MyYellow, label='Analytical Solution')

# Plot the initial approximation
plt.plot(mu0[0], mu0[1], 'd', color=MyRed, linewidth=1,
         markersize=7, markerfacecolor=MyRed, label='Initial Approximation')

# Plot CE iterations
for j in range(CEstr['iter']):
    if j == 0:
        plt.plot(CEstr['xmean'][j, 0], CEstr['xmean'][j, 1], 'd',
                 color=MyRed, linewidth=1, markersize=7, label='CE Iterations')
    else:
        plt.plot(CEstr['xmean'][j, 0], CEstr['xmean'][j, 1], 'd',
                 color=MyRed, linewidth=1, markersize=7, label='_nolegend_')

# Plot the CE solution
plt.plot(Xopt[0, 0], Xopt[0, 1], 'x', color=MyBlue, linewidth=3,
         markersize=25, label='CE Optimum')

# Labeling
plt.xlabel('x1', fontsize=20)
plt.ylabel('x2', fontsize=20)

# Legend (placed in the upper right, equivalent to Matlab's 'NorthEast')
plt.legend(loc='upper right', fontsize=12)

# Set additional font and box properties
ax = plt.gca()
for label in ax.get_xticklabels():
    label.set_fontname('Arial')
for label in ax.get_yticklabels():
    label.set_fontname('Arial')
ax.tick_params(labelsize=18)
# Box on
ax.set_frame_on(True)

# Save the figure as PNG
plt.savefig('CEoptExample2.png', format='png', dpi=300)

# Show the figure
plt.show()
